libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -101,21 +101,13 @@ string IRGenerator::generate(ContractDefinition const& _contract)
 	t("memoryInit", memoryInit());
 	t("constructor", constructorCode(_contract));
 	t("deploy", deployCode(_contract));
-	// We generate code for all functions and rely on the optimizer to remove them again
-	// TODO it would probably be better to only generate functions when internalDispatch or
-	// virtualFunctionName is called - same below.
-	for (auto const* contract: _contract.annotation().linearizedBaseContracts)
-		for (auto const* fun: contract->definedFunctions())
-			generateFunction(*fun);
+	collectFunctions(_contract);
 	t("functions", m_context.functionCollector().requestedFunctions());
 
 	resetContext(_contract);
-	m_context.setMostDerivedContract(_contract);
 	t("RuntimeObject", runtimeObjectName(_contract));
 	t("dispatch", dispatchRoutine(_contract));
-	for (auto const* contract: _contract.annotation().linearizedBaseContracts)
-		for (auto const* fun: contract->definedFunctions())
-			generateFunction(*fun);
+	collectFunctions(_contract);
```

Comment:
This gets repeated twice and I don't see why it would generate different code each time (apart from stuff added to the collector by `generateFunction()` calls in `dispatchRoutine()` but this shouldn't affect other functions?) but it does (several tests fail) so after some attempts to have it reuse the same code I gave up and left it as is.

Reply:
It should be different code: In the first one (the runtime code), we seed the queue by the external interface of the function, in the second one, we seed it by the constructor (or the functions called from there).

Reply:
ok. I'll change it then.